src/lib/provable/test/sha2.unit-test.ts

```diff
@@ -0,0 +1,309 @@
+// In-circuit tests of the SHA2 family of hash functions together with
+// test vectors extracted from https://www.di-mgt.com.au/sha_testvectors.html
+
+import { ZkProgram } from '../../proof-system/zkprogram.js';
+import { Bytes } from '../wrapped-classes.js';
+import { Gadgets } from '../gadgets/gadgets.js';
+import {
+  sha256 as nobleSha256,
+  sha224 as nobleSha224,
+} from '@noble/hashes/sha256';
+import {
+  sha384 as nobleSha384,
+  sha512 as nobleSha512,
+} from '@noble/hashes/sha512';
+import { bytes } from './test-utils.js';
+import {
+  equivalentAsync,
+  equivalentProvable,
+} from '../../testing/equivalent.js';
+import { Random, sample } from '../../testing/random.js';
+import { expect } from 'expect';
+
+// SHA2-224 TESTS
+{
+  sample(Random.nat(400), 5).forEach((preimageLength) => {
+    let inputBytes = bytes(preimageLength);
+    let outputBytes = bytes(224 / 8);
+
+    equivalentProvable({ from: [inputBytes], to: outputBytes, verbose: true })(
+      (x) => nobleSha224(x),
+      (x) => Gadgets.SHA2.hash(224, x),
+      `sha224 preimage length ${preimageLength}`
+    );
+  });
+
+  const Sha2_224Program = ZkProgram({
+    name: `sha2-224`,
+    publicOutput: Bytes(28),
+    methods: {
+      sha224: {
+        privateInputs: [Bytes(192)],
+        async method(preImage: Bytes) {
+          return {
+            publicOutput: Gadgets.SHA2.hash(224, preImage),
+          };
+        },
+      },
+    },
+  });
+
+  const RUNS = 2;
+
+  await Sha2_224Program.compile();
+
+  await equivalentAsync(
+    {
+      from: [bytes(192)],
+      to: bytes(28),
+    },
+    { runs: RUNS }
+  )(nobleSha224, async (x) => {
+    const { proof } = await Sha2_224Program.sha224(x);
+    await Sha2_224Program.verify(proof);
+    return proof.publicOutput;
+  });
+
+  for (let { preimage, hash } of testVectors()) {
+    let actual = Gadgets.SHA2.hash(224, Bytes.fromString(preimage));
+    expect(actual.toHex()).toEqual(hash);
+  }
+
+  function testVectors() {
```

Comment:
Can you add a one-line comment to mention where they come from?

Reply:
It is already there, top line of the file 😅 

> // In-circuit tests of the SHA2 family of hash functions together with
// test vectors extracted from https://www.di-mgt.com.au/sha_testvectors.html